### PR TITLE
[quant][fx] Fully align convert with the reference model design and simplify the implementation

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73863

Summary:
This PR fully aligns the convert function with the design: https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
and simplifies the implementation of convert function by always produce a reference quantized model (with reference patterns) first,
and then lower the model to a quantized model that is runnable with PyTorch native backend (fbgemm/qnnpack).

This PR makes the convert.py much easier to understand than the previous implementation, and we are able to remove majority of code
in quantization_patterns.py as well (in followup PRs).

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```
and other internal/oss regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34778506](https://our.internmc.facebook.com/intern/diff/D34778506)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34778506).

File: torch/ao/quantization/fx/_convert_do_not_use.py [hunk 1]
```diff
@@ -55,6 +78,138 @@
...
+    return quantized
+
+def get_module_path_and_prefix(
+        obs_node: Node,
+        node_name_to_scope: Dict[str, Tuple[str, type]],
+        qconfig_map: Dict[str, QConfigAny]):
+    # an observer can be inserted for both input of the next operator or output of the previous
+    # operator (they can be the same)
+    # this flag identifies if the observer is inserted only because the observed node is
+    # the input of the next operator
```
Comment (andrewor14) [inline torch/ao/quantization/fx/_convert_do_not_use.py:182]: Maybe add a docstring like the other functions? Also what is `this_flag` referring to?

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 2]
```diff
@@ -245,6 +280,26 @@ def should_skip_lowering(op: torch.fx.node.Node, qconfig_map: Dict[str, QConfigA
     torch._ops.ops.quantized.conv3d_prepack,
 }
 
+# Mapping from a functional to (activation_compute_dtype, weight_dtype) to lower to a 2-tuple of
+#   1) The dynamically quantized version of the op
+#   2) The dynamically quantized version of the op fused with relu, if it exists, else None
```
Comment (andrewor14) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:285]: Looks like this comment is out of date now? Or maybe the wording is confusing, maybe something like
```
Mapping from a functional to a dictionary, where the key is a 2-tuple of (...)
and the value is a 2-tuple of (...)
```

**Review by andrewor14 — APPROVED** [review]

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:285]: sounds good, yeah I can change the wording, thanks

File: torch/nn/quantized/dynamic/modules/linear.py [hunk 3]
```diff
@@ -110,3 +110,17 @@ def from_float(cls, mod):
...
+        module
+        Args:
+            ref_qlinear (Module): a reference quantized  module, either produced by
+            torch.ao.quantization functions or provided by the user
+        """
+        qlinear = cls(ref_qlinear.in_features, ref_qlinear.out_features)
+        qweight = ref_qlinear.get_quantized_weight()
+        bias = ref_qlinear.bias
+        qlinear.set_weight_bias(qweight, bias)
+        return qlinear
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/nn/quantized/dynamic/modules/linear.py:126]: sounds good, will add a TODO comment for now, we can enable in a separate PR

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34778506).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34778506).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34778506).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34778506).

**Pull request closed without merging**